README.md
# Leaflet Environmental Layers
[![Build Status](https://travis-ci.org/publiclab/leaflet-environmental-layers.svg)](https://travis-ci.org/publiclab/leaflet-environmental-layers)
[![npm version](https://badge.fury.io/js/leaflet-environmental-layers.svg)](https://badge.fury.io/js/leaflet-environmental-layers) [![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com)
[![Code of Conduct](https://img.shields.io/badge/code-of%20conduct-green.svg)](https://publiclab.org/conduct)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
<a href='https://bower.io/search/'><img src="https://benschwarz.github.io/bower-badges/badge@2x.png" width="130" height="30" alt='Bower Version'></a>

## Use cases

The library is integrated to following places

* https://mapknitter.org/
* https://publiclab.org/maps/

## About

The information of each layer can be found here:

https://publiclab.org/notes/sagarpreet/06-06-2018/leaflet-environmental-layer-library

### Layers

| Layer Name                  | Color   |
| --------------------------- | ------- |
| wisconsin                   | N/A     |
| fracTrackerMobile           | N/A     |
| purpleLayer                 | #8b0000 |
| purpleairmarker             | #800080 |
| skyTruth                    | #ff0000 |
| fractracker                 | #ffff00 |
| pfasLayer                   | #00ff00 |
| toxicReleaseLayer           | #008000 |
| odorreport                  | #ff00ff |
| mapknitter                  | #D50039 |
| Power                       | #ffc0cb |
| Telecom                     | #0000ff |
| Petroleum                   | #a52a2a |
| Water                       | #4B0082 |
| income                      | #006400 |
| americanIndian              | #800000 |
| asian                       | #ffa500 |
| black                       | #FFD700 |
| multi                       | #ffc0cb |
| hispanic                    | #DCDCDC |
| nonWhite                    | #808080 |
| white                       | #a52a2a |
| plurality                   | #800000 |
| clouds                      | #80dfff |
| cloudsClassic               | #b3f0ff |
| precipitation               | #00ff55 |
| precipitationClassic        | #00008b |
| rain                        | #8080ff |
| rainClassic                 | #1a1aff |
| snow                        | #80ffe5 |
| pressure                    | #e62e00 |
| pressureContour             | #ff3300 |
| temperature                 | #ff3300 |
| wind                        | #00008b |
| city                        | #b3ffff |
| windrose                    | #008000 |
| Territories                 | #000000 |
| Languages                   | #000000 |
| Treaties                    | #000000 |
| aqicnLayer                  | #000000 |
| openaq                      | #000000 |
| luftdaten                   | #000000 |
| opensense                   | N/A     |
| osmLandfillMineQuarryLayer  | N/A     |
| eonetFiresLayer             | #78fffa |


## Quick Setup :

### Installation Instructions:
1. Clone this repository to your local environment.
2. Run `npm install` to install all the necessary packages required.
3. Open `examples/index.html` in your browser to look at the preview of the library.

### Instructions for a developer:
1. Install grunt - https://gruntjs.com/installing-grunt.
2. Make the changes you are working on in respective /src files.
3. Run `grunt build` to generate files in the /dist directory.
4. Run `grunt transpile` to transpile es6 code and copy files needed to run the tests to the /dist directory.
5. Run `grunt jasmine` to run tests on the LEL layers and ensure they pass.
6. Test your changes on a browser by opening `examples/index.html`.

## Demo :

Checkout this demo : https://publiclab.github.io/leaflet-environmental-layers/example/index.html#lat=43.00&lon=-4.07&zoom=3&layers=Standard

## Features :

#### Zoom or Pan

Click and drag the map to pan it.

#### Change the Base Map and Overlay layers

Use the button on right-most corner to change the way the background of the map looks .

#### See More Data

Toggle certain layers on and off using the Layers button in the toolbar .

#### Click on a Point

Click on a point or marker on the map to learn more about it .

#### Add a legend

In `src/legendCreation.js`, add `addLayerNameURLPair(layer_var, "img_url");`, where `layer_var` is consistent with the variable used in `example/index.html` and `img_url` is the source of the image to be used as the legend.

#### Add the layers browser menu from the demo page
To add the layers browser menu from the demo page,

Dependencies:
- Requires Bootstrap and jquery to be installed

```
  <!-- Bootstrap --> 
  <link rel="stylesheet" href="../node_modules\bootstrap\dist\css\bootstrap.min.css">
  <script src="../node_modules\jquery\dist\jquery.min.js"></script>
  <script src="../node_modules\bootstrap\dist\js\bootstrap.min.js"></script>
```

```js
  var baseMaps = {
  'Standard': baselayer1,
  };

  var overlayMaps = {
  'Wisconsin Non-Metal': Wisconsin_NM,
  'Indigenous Lands': {
    category: 'group', // Let's the control know if this should be rendered as a group
    layers: { // Layers making the group
      'Territories': IndigenousLandsTerritories,
      'Languages': IndigenousLandsLanguages,
      'Treaties': IndigenousLandsTreaties,
    },
  },
};

  var leafletControl = new L.control.layersBrowser(baseMaps, overlayMaps);
  leafletControl.addTo(map);
```
- `baseMaps` and `overlayMaps` are object literals that have layer names for keys and map layer objects for values.
- `baseMaps` will be hidden if only one base map is provided
- The layer information displayed for each layer is stored in `layerData.json`
- The layer name(key) in the `overlayMaps` object is not case-sensitive and can have spaces but the characters should match with those in `layerData.json`
- The layers are filtered according to the map view
- When there are new layers present in the map view when moving around a badge is displayed near the layer control icon on the top right showing the number of new layers in the view

#### Dependencies for search control

    <script src="../node_modules/leaflet-google-places-autocomplete/src/js/leaflet-gplaces-autocomplete.js"></script>
    <link rel="stylesheet" href="../node_modules/leaflet-google-places-autocomplete/src/css/leaflet-gplaces-autocomplete.css">
    <script src="https://maps.googleapis.com/maps/api/js?key=YOUR_API_KEY&libraries=places"></script>
  
### Add an embed control for embedding the map in other pages :

Add the following code after you have the map(an instance of L.Map) initialized:

    // Assuming your map instance is in a variable called map
    var embedControl = new L.control.embed(options);
    embedControl.addTo(map);

The optional options object can be passed in with any of the following properties:
- position<String>
  The position defaults to 'topleft'.  Other possible values include 'topright', 'bottomleft' or 'bottomright'
- hostname<String>
  Defaults to 'publiclab.github.io'

### Add minimal mode control

    // Assuming your map instance is in a variable called map
    // Assuming your leaflet layers control instance is in a variable called layersControl
    var modeControl = new L.control.minimalMode(layersControl);
    modeControl.addTo(map);

### Spreadsheet-based layers

We can source locations from a spreadsheet in a format like this:

| Title  | Latitude | Longitude   | Notes             |
|--------|----------|-------------|-------------------|
| First	 | 29.671282 | -95.17829  | The first marker  |
| Second | 29.760371 | -95.504828 | The second marker |
| Third  | 29.917755 | -95.283494 | The third marker  |

The layer is constructed like this:

```js
var layer = L.SpreadsheetLayer({
  url: 'https://docs.google.com/spreadsheets/d/14BvU3mEqvI8moLp0vANc7jeEvb0mnmYvH4I0GkwVsiU/edit?usp=sharing', // String url of data sheet
  lat: 'Latitude', // name of latitude column
  lon: 'Longitude', // name of longitude column
  columns: ['Title', 'Notes'], // Array of column names to be used
  generatePopup: function() {
    // function used to create content of popups
  },
  // imageOptions: // optional, defaults to blank
  // sheetNum: // optional, defaults to 0 (first sheet)
});
layer.addTo(map);
```

Read more here: https://github.com/publiclab/leaflet-environmental-layers/blob/master/src/util/googleSpreadsheetLayer.js

We're going to try spinning this out into its own library; see: https://github.com/publiclab/leaflet-environmental-layers/issues/121

## Dependencies :

### General (required for all layers) :

           <script src="../dist/LeafletEnvironmentalLayers.js"></script>
           <link href="../node_modules/leaflet/dist/leaflet.css" rel="stylesheet">
	   <script src="../node_modules/leaflet-spin/example/spin/dist/spin.min.js"></script>  <!-- Compulsory to add -->
 	   <script src="../node_modules/leaflet-spin/example/leaflet.spin.min.js"></script>

### To use Wisconsin Non-Metallic Layer :

            <script src="https://unpkg.com/esri-leaflet@2.2.3/dist/esri-leaflet.js"></script>
            <script src="https://unpkg.com/esri-leaflet-renderers@2.0.6"></script>

              var Wisconsin_NM = wisconsinLayer(map) ;

### To use Fractracker Mobile Layer :
            var FracTracker_mobile = L.geoJSON.fracTrackerMobile();


### To use Purple Layer :

			<script src="../node_modules/heatmap.js/build/heatmap.min.js"></script>
			<script src="../node_modules/leaflet-heatmap/leaflet-heatmap.js"></script>


## Real Time Layers :

1.) city (by openWeather)

        var city = L.OWM.current({intervall: 15, minZoom: 3});

2.) WindRose (by openWeather)
      
      <script src="../src/windRoseLayer.js"></script>
       <link href="../dist/LeafletEnvironmentalLayers.css" rel="stylesheet">

     var windrose = L.OWM.current({intervall: 15, minZoom: 3, markerFunction: myWindroseMarker, popup: false, clusterSize:       50,imageLoadingBgUrl: 'https://openweathermap.org/img/w0/iwind.png' });
    windrose.on('owmlayeradd', windroseAdded, windrose);

## Open Infra Map :

##### OpenInfraMap_Power Layer :

    var OpenInfraMap_Power = L.tileLayer('https://tiles-{s}.openinframap.org/power/{z}/{x}/{y}.png',{
        maxZoom: 18,
        attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>'
    });

##### OpenInfraMap_Petroleum Layer :

    var OpenInfraMap_Petroleum = L.tileLayer('https://tiles-{s}.openinframap.org/petroleum/{z}/{x}/{y}.png', {
      maxZoom: 18,
      attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>'
    });

##### OpenInfraMap_Telecom Layer :

    var OpenInfraMap_Telecom = L.tileLayer('https://tiles-{s}.openinframap.org/telecoms/{z}/{x}/{y}.png', {
      maxZoom: 18,
      attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>'
    });

##### OpenInfraMap_Water Layer :

    var OpenInfraMap_Water = L.tileLayer('https://tiles-{s}.openinframap.org/water/{z}/{x}/{y}.png',{
      maxZoom: 18,
      attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>'
    });   

## Add hash support for easy sharing of map :    

### Steps To add :

1.) Add link :

       <script src="../lib/leaflet-fullUrlHash.js"></script>

2.) Once you have initialized the map (an instance of L.Map), add the following code :

     // Assuming your map instance is in a variable called map and an object with all the map layers is
    in a variable called allMapLayers
    var hash = new L.FullHash(map, allMapLayers);    


 ## Add all LEL Layers at once:

 	 L.LayerGroup.EnvironmentalLayers().addTo(map);

## Add all layers except some layers:

	 L.LayerGroup.EnvironmentalLayers({
            exclude: ['mapknitter', 'clouds'],
         }).addTo(map);

## Add some layers only:

	 L.LayerGroup.EnvironmentalLayers({
            include: ['mapknitter', 'clouds'],
         }).addTo(map);

  The layers added to the 'include' option are displayed by default when the map is initialized

## Add some layers and display them default:

	 L.LayerGroup.EnvironmentalLayers({
            addLayersToMap: true,    // by default this is FALSE
            include: ['mapknitter', 'clouds'],
         }).addTo(map);

## Add base layers:

    L.LayerGroup.EnvironmentalLayers({
            baseLayers: {               // Grayscale base map is used by default when this is undefined 
             'Standard': baselayer1
            },
          }).addTo(map);

## Turn on Leaflet Hash in the URL:

	 L.LayerGroup.EnvironmentalLayers({
            exclude: ['mapknitter', 'clouds'],
	    hash: true,             // by default this is FALSE
         }).addTo(map);

## Turn on embed control:

    L.LayerGroup.EnvironmentalLayers({
          exclude: ['mapknitter', 'clouds'],
          hash: true,                   // by default this is FALSE
          embed: true,                  // by default this is FALSE
          hostname: 'placeholder.org'   // by default this is set to 'publiclab.github.io'
    }).addTo(map);

  By adding a hostname, the embed code can point to the domain that hosts the maps.


## Browse layers

  The layer menu used by default groups similar layers, displays more information about the layers and filters them when out of bounds.

  ### Dependencies
      
        <link rel="stylesheet" href="../node_modules\bootstrap\dist\css\bootstrap.min.css">
        <script src="../node_modules\jquery\dist\jquery.min.js"></script>
        <script src="../node_modules\bootstrap\dist\js\bootstrap.min.js"></script>
        <link href="../dist/LeafletEnvironmentalLayers.css" rel="stylesheet" />
    
  ### To use Leaflet's default layer control:

    L.LayerGroup.EnvironmentalLayers({
            simpleLayerControl: true,   // by default this is FALSE
      }).addTo(map);

CHANGELOG.md
# Changelog
## [Unreleased]

## Latest - 2020-02-01

### Added
- Add a `display` option to the basic usage to display specific layers by default
- Minimal mode markers feature

### Fixed
- Layer menu removing layers from the layer menu showing only globally available layers upon refresh

## 2.2.0 - 2020-01-28

### Added
- Add PL people layer
- Add an `addLayersToMap` option in basic usage
- Ability to choose between using leaflet's default layer control and LEL's new layer menu when instantiating LEL

### Changed
- Bump LBL to 1.7.0
- Implement new layer menu in LEL's basic usage
- Update documentation
- Move new layer menu styles and custom leaflet control styles to dist/LeafletEnvironmentalLayers.css from example/styles.css

### Fixed
- Fix multiple maps sharing the same default base tile layer by [@z-arnott](https://github.com/z-arnott)
- Fix errors on some layers when trying to display them from hash when out of bounds
- Fix styles for layer group name and layer description
- Fix overriding of default leaflet classes


example/layers.js
var bounds = new L.LatLngBounds(
  new L.LatLng(84.67351257, -172.96875),
  new L.LatLng(-54.36775852, 178.59375),
);
var map = L.map('map', {
  maxBounds: bounds,
  maxBoundsViscosity: 0.75,
}).setView([43, -83], 3);
map.options.minZoom = 3;
var baselayer1 = L.tileLayer(
  'https://api.tiles.mapbox.com/v4/mapbox.emerald/{z}/{x}/{y}.png?access_token=pk.eyJ1Ijoic2FnYXJwcmVldDk3IiwiYSI6ImNqaXhlZjgwaDJtb3EzcW1zdDdwMzJkODcifQ.MA2YIv6VpGLLAo-QYUudTA',
  {
    attribution:
            '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  },
).addTo(map);
var baselayer3 = L.tileLayer(
  'https://api.tiles.mapbox.com/v4/mapbox.streets-basic/{z}/{x}/{y}.png?access_token=pk.eyJ1Ijoic2FnYXJwcmVldDk3IiwiYSI6ImNqaXhlZjgwaDJtb3EzcW1zdDdwMzJkODcifQ.MA2YIv6VpGLLAo-QYUudTA',
  {
    attribution:
            '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  },
);
var baselayer2 = L.tileLayer(
  'https://a.tiles.mapbox.com/v3/jywarren.map-lmrwb2em/{z}/{x}/{y}.png',
  {
    attribution:
            '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  },
);
var baselayer4 = L.tileLayer(
  'https://api.tiles.mapbox.com/v4/mapbox.dark/{z}/{x}/{y}.png?access_token=pk.eyJ1Ijoic2FnYXJwcmVldDk3IiwiYSI6ImNqaXhlZjgwaDJtb3EzcW1zdDdwMzJkODcifQ.MA2YIv6VpGLLAo-QYUudTA',
  {
    attribution:
            '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
  },
);

var PLpeople = L.layerGroup.PLpeople();
var PurpleLayer = L.layerGroup.purpleLayer();
var ToxicRelease = L.layerGroup.toxicReleaseLayer();
var PFASTracker = L.layerGroup.pfasLayer();
var AQICNLayer = L.layerGroup.aqicnLayer();
var OSMLandfillMineQuarryLayer = L.layerGroup.osmLandfillMineQuarryLayer();

var PurpleAirMarkerLayer = L.layerGroup.layerCode('purpleairmarker');
var Fractracker = L.layerGroup.layerCode('fractracker');
var SkyTruth = L.layerGroup.layerCode('skytruth');
var OdorReport = L.layerGroup.layerCode('odorreport');
var MapKnitter = L.layerGroup.layerCode('mapknitter');
var OpenAqLayer = L.layerGroup.layerCode('openaq');
var LuftdatenLayer = L.layerGroup.layerCode('luftdaten');
var OpenSenseLayer = L.layerGroup.layerCode('opensense');


var OpenInfraMap_Power = L.tileLayer('https://tiles-{s}.openinframap.org/power/{z}/{x}/{y}.png', {
  maxZoom: 18,
  attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>',
});
var OpenInfraMap_Petroleum = L.tileLayer('https://tiles-{s}.openinframap.org/petroleum/{z}/{x}/{y}.png', {
  maxZoom: 18,
  attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>',
});
var OpenInfraMap_Telecom = L.tileLayer('https://tiles-{s}.openinframap.org/telecoms/{z}/{x}/{y}.png', {
  maxZoom: 18,
  attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>',
});
var OpenInfraMap_Water = L.tileLayer('https://tiles-{s}.openinframap.org/water/{z}/{x}/{y}.png', {
  maxZoom: 18,
  attribution: '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a>, <a href="http://www.openinframap.org/about.html">About OpenInfraMap</a>',
});

var Justicemap_income = L.tileLayer.provider('JusticeMap.income');
var JusticeMap_americanIndian = L.tileLayer.provider('JusticeMap.americanIndian');
var JusticeMap_asian = L.tileLayer.provider('JusticeMap.asian');
var JusticeMap_black = L.tileLayer.provider('JusticeMap.black');
var JusticeMap_multi = L.tileLayer.provider('JusticeMap.multi');
var JusticeMap_hispanic = L.tileLayer.provider('JusticeMap.hispanic');
var JusticeMap_nonWhite = L.tileLayer.provider('JusticeMap.nonWhite');
var JusticeMap_white = L.tileLayer.provider('JusticeMap.white');
var JusticeMap_plurality = L.tileLayer.provider('JusticeMap.plurality');

var clouds = L.OWM.clouds({showLegend: true, opacity: 0.5});
var cloudscls = L.OWM.cloudsClassic({});
var precipitation = L.OWM.precipitation({});
var precipitationcls = L.OWM.precipitationClassic({});
var rain = L.OWM.rain({});
var raincls = L.OWM.rainClassic({});
var snow = L.OWM.snow({});
var pressure = L.OWM.pressure({});
var pressurecntr = L.OWM.pressureContour({});
var temp = L.OWM.temperature({});
var wind = L.OWM.wind({});


var city = L.OWM.current({intervall: 15, minZoom: 3});
var windrose = L.OWM.current({
  intervall: 15,
  minZoom: 3,
  markerFunction: myWindroseMarker,
  popup: false,
  clusterSize: 50,
  imageLoadingBgUrl: 'https://openweathermap.org/img/w0/iwind.png',
});
windrose.on('owmlayeradd', windroseAdded, windrose);

var IndigenousLandsTerritories = L.layerGroup.indigenousLayers('Territories');
var IndigenousLandsLanguages = L.layerGroup.indigenousLayers('Languages');
var IndigenousLandsTreaties = L.layerGroup.indigenousLayers('Treaties');
var Unearthing = L.layerGroup.Unearthing();

var Wisconsin_NM = wisconsinLayer(map);
var FracTracker_mobile = L.geoJSON.fracTrackerMobile();

var EonetFiresLayer = L.geoJSON.eonetFiresLayer();

var baseMaps = {
  'Standard': baselayer1,
  // 'Grey-scale': baselayer2,
  // 'Streets': baselayer3,
  // 'Dark': baselayer4,
};

var overlayMaps = {
  "PLpeople" : PLpeople,
  'wisconsin': Wisconsin_NM,
  'fractracker': Fractracker,
  'fracTrackerMobile': FracTracker_mobile,
  'purpleair': {
    category: 'group',
    layers: {
      'PurpleAirLayer-HeatMap': PurpleLayer,
      'PurpleAirMarkerLayer': PurpleAirMarkerLayer,
    }
  },
  'skytruth': SkyTruth,
  'toxicReleaseLayer': ToxicRelease,
  'odorreport': OdorReport,
  'mapknitter': MapKnitter,
  'openInfraMap': {
    category: 'group',
    layers: {
      'Power': OpenInfraMap_Power,
      'Telecom': OpenInfraMap_Telecom,
      'Petroleum': OpenInfraMap_Petroleum,
      'Water': OpenInfraMap_Water,
    }
  },
  'indigenousLands': {
    category: 'group',
    layers: {
      'Territories': IndigenousLandsTerritories,
      'Languages': IndigenousLandsLanguages,
      'Treaties': IndigenousLandsTreaties,
    },
  },
  'justiceMap': {
    category: 'group',
    layers: {
      'income': Justicemap_income,
      'americanIndian': JusticeMap_americanIndian,
      'asian': JusticeMap_asian,
      'black': JusticeMap_black,
      'multi': JusticeMap_multi,
      'hispanic': JusticeMap_hispanic,
      'nonWhite': JusticeMap_nonWhite,
      'white': JusticeMap_white,
      'plurality': JusticeMap_plurality,
    },
  },
  'openWeatherMap': {
    category: 'group',
    layers: {
      'Clouds': clouds,
      'clouds (classic)': cloudscls,
      'precipitation': precipitation,
      'precipitation (classic)': precipitationcls,
      'rain': rain,
      'rain (classic)': raincls,
      'snow': snow,
      'pressure': pressure,
      'pressure contour (zoom in)': pressurecntr,
      'temp': temp,
      'wind': wind,
      'Cities (zoom in)': city,
      'windrose (zoom in)': windrose,
    },
  },
  'pfaslayer': PFASTracker,
  
  'aqicnLayer': AQICNLayer,
  'openaq': OpenAqLayer,
  'luftdaten': LuftdatenLayer,
  'opensense': OpenSenseLayer,
  'osmLandfillMineQuarryLayer': OSMLandfillMineQuarryLayer,
  'eonetFiresLayer': EonetFiresLayer,
  'unearthing': Unearthing
};

var allMapLayers = {
  'Standard': baselayer1,
  'Grey-scale': baselayer2,
  'Streets': baselayer3,
  'Dark': baselayer4,

  "PLpeople" : PLpeople,
  'wisconsin': Wisconsin_NM,
  'fracTrackerMobile': FracTracker_mobile,
  'purpleLayer': PurpleLayer,
  'purpleairmarker': PurpleAirMarkerLayer,
  'skytruth': SkyTruth,
  'fractracker': Fractracker,
  'pfasLayer': PFASTracker,
  'toxicReleaseLayer': ToxicRelease,
  'odorreport': OdorReport,
  'mapknitter': MapKnitter,
  'Power': OpenInfraMap_Power,
  'Telecom': OpenInfraMap_Telecom,
  'Petroleum': OpenInfraMap_Petroleum,
  'Water': OpenInfraMap_Water,
  'income': Justicemap_income,
  'americanIndian': JusticeMap_americanIndian,
  'asian': JusticeMap_asian,
  'black': JusticeMap_black,
  'multi': JusticeMap_multi,
  'hispanic': JusticeMap_hispanic,
  'nonWhite': JusticeMap_nonWhite,
  'white': JusticeMap_white,
  'plurality': JusticeMap_plurality,
  'clouds': clouds,
  'cloudsClassic': cloudscls,
  'precipitation': precipitation,
  'precipitationClassic': precipitationcls,
  'rain': rain,
  'rainClassic': raincls,
  'snow': snow,
  'pressure': pressure,
  'pressureContour': pressurecntr,
  'temperature': temp,
  'wind': wind,
  'city': city,
  'windrose': windrose,
  'Territories': IndigenousLandsTerritories,
  'Languages': IndigenousLandsLanguages,
  'Treaties': IndigenousLandsTreaties,
  'aqicnLayer': AQICNLayer,
  'openaq': OpenAqLayer,
  'luftdaten': LuftdatenLayer,
  'opensense': OpenSenseLayer,
  'osmLandfillMineQuarryLayer': OSMLandfillMineQuarryLayer,
  'eonetFiresLayer': EonetFiresLayer,
  'Unearthing': Unearthing
};

// var oms = omsUtil(map, {
//   keepSpiderfied: true,
//   circleSpiralSwitchover: 0
// });

var hash = new L.FullHash(map, allMapLayers);
// var leafletControl = new L.control.layers(baseMaps, overlayMaps);
var leafletControl = new L.control.layersBrowser(baseMaps, overlayMaps);
leafletControl.addTo(map);

var modeControl = new L.control.minimalMode(leafletControl);
modeControl.addTo(map);

var embedControl = new L.control.embed({
  // hostname: 'your domain name goes here'
});
embedControl.addTo(map);

// Collapsible search control
new L.Control.GPlaceAutocomplete({
  position: 'topleft',
  collapsed_mode: true,
	callback: function(place){
		var loc = place.geometry.location;
		map.setView( [loc.lat(), loc.lng()], 18);
	}
}).addTo(map);
